### PR TITLE
Make content-type backend serialization pluggable

### PR DESCRIPTION
**Pull request by snazy**
This also removes the necessity to instantiate a `StoreWorker`,
implicitly cleaning up a bunch of related code.

Also finally removes the technical ability to store Iceberg content
with global state.

This PR is split into two commits: one with the actual changes and another with the "boring stuff/changes" implied by the actual changes.

Depends on PRs: #4941 #4942 #4943

Follow-up PRs: #4945

See issue #4940

### CODE REVIEW

File: versioned/spi/src/main/java/org/projectnessie/versioned/store/ContentSerializer.java [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+import org.projectnessie.model.Content;
+import org.projectnessie.versioned.ContentAttachment;
+import org.projectnessie.versioned.ContentAttachmentKey;
+
+/**
+ * Content serializers provide persistence layer (de)serialization functionality for a specific
+ * content type.
+ */
+public interface ContentSerializer<C extends Content> {
+  String contentType();
```
Comment (dimas-b) [inline versioned/spi/src/main/java/org/projectnessie/versioned/store/ContentSerializer.java:32]: Why not `ContentType` object?

File: versioned/spi/src/main/java/org/projectnessie/versioned/store/DefaultStoreWorker.java [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+    return Lazy.INSTANCE;
+  }
+
+  private static final class Lazy {
+    private static final DefaultStoreWorker INSTANCE = new DefaultStoreWorker();
+  }
+
+  private static final class Registry {
+    private static final ContentSerializer<?>[] BY_PAYLOAD;
+    private static final Map<Class<? extends Content>, ContentSerializer<?>> BY_TYPE;
```
Comment (dimas-b) [inline versioned/spi/src/main/java/org/projectnessie/versioned/store/DefaultStoreWorker.java:53]: Why not `Map<ContentType, ContentSerializer<?>>`?

File: versioned/spi/src/main/java/org/projectnessie/versioned/store/ContentSerializer.java [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+  C applyId(C content, String id);
+
+  default boolean requiresGlobalState(C content) {
+    return false;
+  }
+
+  Content.Type getType(byte payload, ByteString onReferenceValue);
+
+  boolean requiresGlobalState(byte payload, ByteString onReferenceValue);
```
Comment (dimas-b) [inline versioned/spi/src/main/java/org/projectnessie/versioned/store/ContentSerializer.java:44]: Could you move it one method up to be together with `requiresGlobalState(C content)`?

File: servers/store/src/main/java/org/projectnessie/server/store/IcebergTableSerializer.java [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+public final class IcebergTableSerializer extends BaseSerializer<IcebergTable> {
+
+  @Override
+  public String contentType() {
+    return Content.Type.ICEBERG_TABLE.name();
+  }
+
+  @Override
+  public Content.Type getType(byte payload, ByteString onReferenceValue) {
+    return Content.Type.ICEBERG_TABLE;
```
Comment (dimas-b) [inline servers/store/src/main/java/org/projectnessie/server/store/IcebergTableSerializer.java:33]: What is the purpose of the args now? If only for test, could we validate payload to value match some other way, without overloading the main production call path with extra args?

Comment (dimas-b) [reply servers/store/src/main/java/org/projectnessie/server/store/IcebergTableSerializer.java]: Ah, thx, I missed that... However, would it make sense to rename `UnknownSerializer` to `LegacySerializer` perhaps?.. just for the sake of clarity... I'd even call it `LegacyDeerializer`.

Comment (snazy) [reply servers/store/src/main/java/org/projectnessie/server/store/IcebergTableSerializer.java]: The type‘s named `UNKNOWN`, that’s why

Comment (dimas-b) [reply servers/store/src/main/java/org/projectnessie/server/store/IcebergTableSerializer.java]: But them we have type misalignment between `contentType()` and `getType(byte payload, ByteString onReferenceValue)`.

`contentType()` appears to be required only for serialization, which is not supported by `UnknownSerializer`... looks like of odd.

What do you think about making the `0` payload handling a special case in `DefaultStoreWorker`? Then, we may not need `UnknownSerializer` at all :thinking:

Comment (snazy) [reply servers/store/src/main/java/org/projectnessie/server/store/IcebergTableSerializer.java]: `DefaultStoreWorker` shouldn't know anything about the "main" content types.
We can likely remove the functionality in `UnknownSerializer` when there are no more commits with payload==0 - the state where we'll eventually be.

Comment (snazy) [reply servers/store/src/main/java/org/projectnessie/server/store/IcebergTableSerializer.java]: I.e. `UnknownSerializer` doesn't need to be perfect - it'll go away.

Comment (snazy) [reply servers/store/src/main/java/org/projectnessie/server/store/IcebergTableSerializer.java]: ... and the `getType()` function will then disappear as well

Comment (dimas-b) [reply servers/store/src/main/java/org/projectnessie/server/store/IcebergTableSerializer.java]: yes, but it still looks awkward. How about using a separate type for `0` payload, i.e. `LEGACY`, which would _not_ have a constant in `Content.Type`. The `UNKNOWN` type's serialize would throw exceptions on ser/deser attempts and the `LEGACY` serializer would be used only for reading old entries with `0` payload. WDYT?

Comment (dimas-b) [reply servers/store/src/main/java/org/projectnessie/server/store/IcebergTableSerializer.java]: Actually, we can probably resolve this after #4945

Comment (snazy) [reply servers/store/src/main/java/org/projectnessie/server/store/IcebergTableSerializer.java]: `UNKNOWN` itself is useful - i.e. when there's a payload for which you no longer have a serializer/content.

it's just that accidentally existing persisted data has payload==0, which should have never happened, but probably did.

Comment (dimas-b) [reply servers/store/src/main/java/org/projectnessie/server/store/IcebergTableSerializer.java]: Actually, we can probably resolve this after #4945

**Review by dimas-b — APPROVED** [review]

**Pull request merged**